egs2/TEMPLATE/asr1/asr.sh

```diff
@@ -330,185 +332,322 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
             --model_prefix="${bpeprefix}" \
             --input_sentence_size="${bpe_input_sentence_size}"
 
-        echo "${oov}" > "${token_list}"
-        # shellcheck disable=SC2002
-        cat ${srctexts} | cut -f 2- -d" " | \
-            spm_encode --model="${bpemodel}" --output_format=piece | \
-                tr ' ' '\n' | sort -u >> "${token_list}"
+        _opts="--bpemodel ${bpemodel}"
 
     elif [ "${token_type}" = char ]; then
         log "Stage 3: Generate character level token_list from ${srctexts}"
-        mkdir -p "$(dirname ${token_list})"
+        _opts="--non_linguistic_symbols ${nlsyms_txt}"
 
-
-        echo "${oov}" > "${token_list}"
-        if [ -n "${nlsyms}" ]; then
-            # shellcheck disable=SC2002
-            cat ${srctexts} | pyscripts/text/text2token.py -s 1 -n 1 -l "${nlsyms}"  \
-                | cut -f 2- -d" " | tr " " "\n" | sort -u \
-                | grep -v -e '^\s*$' >> "${token_list}"
-        else
-            # shellcheck disable=SC2002
-            cat ${srctexts} | pyscripts/text/text2token.py -s 1 -n 1 \
-                | cut -f 2- -d" " | tr " " "\n" | sort -u \
-                | grep -v -e '^\s*$' >> "${token_list}"
-        fi
     else
         log "Error: not supported --token_type '${token_type}'"
         exit 2
     fi
 
+    mkdir -p "$(dirname ${token_list})"
+
+    # The first symbol in token_list must be "<blank>" and the last must be also sos/eos:
+    # 0 is reserved for CTC-blank for ASR and also used as ignore-index in the other task
+
+    # shellcheck disable=SC2002
+    cat ${srctexts} | \
+        python3 -m espnet2.bin.tokenize_text  \
+            --token_type "${token_type}" -f 2- --input - --output - ${_opts} \
+            --field 2- \
+            --write_vocabulary true \
+            --add_symbol "${blank}:0" \
+            --add_symbol "${oov}:1" \
+            --add_symbol "${sos_eos}:-1" \
+        > "${token_list}"
+
```

Comment:
I removed `echo ${blank}...` and `cut -f` and and embed  `espnet2.bin.tokenize_text`.